src/modules/emote_autocomplete/twitch/EmoteAutocomplete.jsx

```diff
@@ -0,0 +1,146 @@
+import {EmoteCategories} from '../../../constants.js';
+import './EmoteAutocomplete.module.css';
+
+const EMOTE_ID_BETTERTTV_PREFIX = '__BTTV__';
+const CUSTOM_SET_ID = 'BETTERTTV_EMOTES';
+const AUTOCOMPLETE_MATCH_IMAGE_QUERY = '.emote-autocomplete-provider__image, .chat-line__message--emote';
+
+const PATCHED_SENTINEL = Symbol('patched symbol');
+
+function serializeEmoteId(emote) {
+  const data = `${emote.category.provider}-${emote.id}-${emote.code}`;
+  return `${EMOTE_ID_BETTERTTV_PREFIX}${encodeURIComponent(btoa(data))}`;
+}
+
+function deserializeEmoteFromURL(url) {
+  const emoteData = url.split(EMOTE_ID_BETTERTTV_PREFIX)[1]?.split('/')[0];
+  if (emoteData == null) {
+    return null;
+  }
+  try {
+    const [emoteProvider, emoteId, emoteCode] = atob(decodeURIComponent(emoteData)).split('-');
+    return {
+      provider: emoteProvider,
+      id: emoteId,
+      code: emoteCode,
+    };
+  } catch (_) {
+    return null;
+  }
+}
+
+function getAutocompleteEmoteProvider() {
+  const autocompleteNode = twitch.getAutocompleteStateNode();
+  if (autocompleteNode == null) {
+    return null;
+  }
+
+  const autocompleteProviders = autocompleteNode.stateNode.providers;
+  return autocompleteProviders.find(({autocompleteType}) => autocompleteType === 'emote');
+}
+
+function createTwitchEmoteSet(allEmotes) {
+  return {
+    emotes: allEmotes.map((emote) => ({
+      id: serializeEmoteId(emote),
+      modifiers: null,
+      setID: CUSTOM_SET_ID,
+      token: emote.code,
+    })),
+    id: CUSTOM_SET_ID,
+  };
+}
+
+function injectEmoteSets() {
+  const autocompleteEmoteProvider = getAutocompleteEmoteProvider();
+  if (autocompleteEmoteProvider == null) {
+    return;
+  }
+
+  const autocompleteEmotes = autocompleteEmoteProvider.props.emotes;
+
+  const allEmotes = emotes.getEmotesByCategories([
+    EmoteCategories.BETTERTTV_CHANNEL,
+    EmoteCategories.BETTERTTV_GLOBAL,
+    EmoteCategories.BETTERTTV_PERSONAL,
+    EmoteCategories.FRANKERFACEZ_CHANNEL,
+    EmoteCategories.FRANKERFACEZ_GLOBAL,
+  ]);
+
+  const emoteSet = createTwitchEmoteSet(allEmotes);
+  const index = autocompleteEmotes.findIndex(({id}) => id === CUSTOM_SET_ID);
+
+  if (index === -1) {
+    autocompleteEmotes.push(emoteSet);
+  } else {
+    autocompleteEmotes[index] = emoteSet;
+  }
+}
+
+function patchEmoteImage(image, isConnected) {
+  if (!isConnected) {
+    return;
+  }
+
+  const url = String(image.srcset);
+  const deseralizedEmote = deserializeEmoteFromURL(url);
+
+  if (deseralizedEmote == null) {
+    return;
+  }
+
+  const emote = emotes.getEligibleEmote(deseralizedEmote.code);
+  if (emote == null) {
+    return;
+  }
+
+  image.srcset = createSrcSet(emote.images);
+  image.src = emote.images['1x'];
+}
+
+let twitchComponentDidUpdate = null;
+export default class EmoteAutocomplete {
+  constructor() {
+    this.load();
+    watcher.on('channel.updated', () => this.load());
+
+    emoteMenuViewStore.on('dirty', emoteMenuViewStore.updateEmotes);
```

Comment:
why would we always update the emotes when dirty? this defeats the whole purpose of this dirty optimization, no?

this also errors:

![image](https://user-images.githubusercontent.com/2942184/161438628-ba1856b7-f610-40fb-bc0a-f929a4f51da4.png)